src/thor/bidirectional_astar.cc

```diff
@@ -202,10 +198,96 @@ void BidirectionalAStar::ExpandForward(GraphReader& graphreader,
   }
 }
 
+// Runs in the inner loop of `ExpandForward`, essentially evaluating if
+// the edge described in `meta` should be placed on the stack
+// as well as doing just that.
+//
+// TODO: Merge this with ExpandReverseInner
+//
+// Returns true if the edge was added
+bool BidirectionalAStar::ExpandForwardInner(GraphReader& graphreader,
+                                            const BDEdgeLabel& pred,
+                                            const NodeInfo* nodeinfo,
+                                            const uint32_t pred_idx,
+                                            const EdgeMetadata& meta,
+                                            uint32_t& shortcuts,
+                                            const GraphTile* tile) {
+  // Skip shortcut edges until we have stopped expanding on the next level. Use regular
+  // edges while still expanding on the next level since we can still transition down to
+  // that level. If using a shortcut, set the shortcuts mask. Skip if this is a regular
+  // edge superseded by a shortcut.
+  if (meta.edge->is_shortcut()) {
+    if (hierarchy_limits_forward_[meta.edge_id.level() + 1].StopExpanding()) {
+      shortcuts |= meta.edge->shortcut();
+    } else {
+      return false;
+    }
+  } else if (shortcuts & meta.edge->superseded()) {
+    return false;
+  }
+
+  // Skip this edge if edge is permanently labeled (best path already found
+  // to this directed edge), if no access is allowed (based on costing method),
+  // or if a complex restriction prevents transition onto this edge.
+  bool is_allowed = costing_->Allowed(meta.edge, pred, tile, meta.edge_id, 0, 0);
+  bool is_restricted =
+      costing_->Restricted(meta.edge, pred, edgelabels_forward_, tile, meta.edge_id, true);
+  if (meta.edge_status->set() == EdgeSet::kPermanent || !is_allowed || is_restricted) {
```

Comment:
Some of the refactor performance drop may come from `is_allowed`/`is_restricted` being evaluated every time instead of using the early return function of ` ... || ... `.

Reply:
ah wow i didnt even notice this, good catch, we should avoid as much as possible extra costing calls